
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>preprocess.raw_data_functions &#8212; easyPheno  documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nature.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">easyPheno  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">preprocess.raw_data_functions</span></code></a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-preprocess.raw_data_functions">
<span id="preprocess-raw-data-functions"></span><h1><a class="reference internal" href="#module-preprocess.raw_data_functions" title="preprocess.raw_data_functions"><code class="xref py py-mod docutils literal notranslate"><span class="pre">preprocess.raw_data_functions</span></code></a><a class="headerlink" href="#module-preprocess.raw_data_functions" title="Permalink to this headline">¶</a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.prepare_data_files" title="preprocess.raw_data_functions.prepare_data_files"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prepare_data_files</span></code></a>(data_dir, genotype_matrix_name, phenotype_matrix_name, phenotype, datasplit, n_outerfolds, n_innerfolds, test_set_size_percentage, val_set_size_percentage, models, user_encoding, maf_percentage)</p></td>
<td><p>Prepare all data files for a common format: genotype matrix, phenotype matrix and index file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_h5_file" title="preprocess.raw_data_functions.check_genotype_h5_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_genotype_h5_file</span></code></a>(data_dir, genotype_matrix_name, encodings)</p></td>
<td><p>Check .h5 genotype file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_index_file" title="preprocess.raw_data_functions.check_index_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_index_file</span></code></a>(data_dir, genotype_matrix_name, phenotype_matrix_name, phenotype)</p></td>
<td><p>Check if index file is available and if the datasets 'y', 'matched_sample_ids', 'X_index', 'y_index' and</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.save_all_data_files" title="preprocess.raw_data_functions.save_all_data_files"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_all_data_files</span></code></a>(data_dir, genotype_matrix_name, phenotype_matrix_name, phenotype, models, user_encoding, maf_percentage, datasplit, n_outerfolds, n_innerfolds, test_set_size_percentage, val_set_size_percentage)</p></td>
<td><p>Prepare and save all required data files:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_transform_format_genotype_matrix" title="preprocess.raw_data_functions.check_transform_format_genotype_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_transform_format_genotype_matrix</span></code></a>(data_dir, genotype_matrix_name, models, user_encoding)</p></td>
<td><p>Check the format of the specified genotype matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_csv_file" title="preprocess.raw_data_functions.check_genotype_csv_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_genotype_csv_file</span></code></a>(data_dir, genotype_matrix_name, encodings)</p></td>
<td><p>Load .csv genotype file. File must have the following structure:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_binary_plink_file" title="preprocess.raw_data_functions.check_genotype_binary_plink_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_genotype_binary_plink_file</span></code></a>(data_dir, genotype_matrix_name)</p></td>
<td><p>Load binary PLINK file, .bim, .fam, .bed files with same prefix need to be in same folder.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_plink_file" title="preprocess.raw_data_functions.check_genotype_plink_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_genotype_plink_file</span></code></a>(data_dir, genotype_matrix_name)</p></td>
<td><p>Load PLINK files, .map and .ped file with same prefix need to be in same folder.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.create_genotype_h5_file" title="preprocess.raw_data_functions.create_genotype_h5_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_genotype_h5_file</span></code></a>(data_dir, genotype_matrix_name, sample_ids, snp_ids, X)</p></td>
<td><p>Save genotype matrix in unified .h5 file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_and_load_phenotype_matrix" title="preprocess.raw_data_functions.check_and_load_phenotype_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_and_load_phenotype_matrix</span></code></a>(data_dir, phenotype_matrix_name, phenotype)</p></td>
<td><p>Check and load the specified phenotype matrix. Only accept .csv, .pheno, .txt files.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.genotype_phenotype_matching" title="preprocess.raw_data_functions.genotype_phenotype_matching"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genotype_phenotype_matching</span></code></a>(X, X_ids, y)</p></td>
<td><p>Match the handed over genotype and phenotype matrix for the phenotype specified by the user</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.get_matched_data" title="preprocess.raw_data_functions.get_matched_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_matched_data</span></code></a>(data, index)</p></td>
<td><p>Get elements of data specified in index array</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.append_index_file" title="preprocess.raw_data_functions.append_index_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append_index_file</span></code></a>(data_dir, genotype_matrix_name, phenotype_matrix_name, phenotype, datasplit, n_outerfolds, n_innerfolds, test_set_size_percentage, val_set_size_percentage, maf_percentage)</p></td>
<td><p>Check index file, described in create_index_file(), and append datasets if necessary</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.create_index_file" title="preprocess.raw_data_functions.create_index_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_index_file</span></code></a>(data_dir, genotype_matrix_name, phenotype_matrix_name, phenotype, datasplit, n_outerfolds, n_innerfolds, test_set_size_percentage, val_set_size_percentage, maf_percentage, X, y, sample_ids, X_index, y_index)</p></td>
<td><p>Create the .h5 index file containing the maf filters and data splits for the combination of genotype matrix,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.filter_non_informative_snps" title="preprocess.raw_data_functions.filter_non_informative_snps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_non_informative_snps</span></code></a>(X)</p></td>
<td><p>Remove non-informative SNPs, i.e. SNPs that are constant</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.get_minor_allele_freq" title="preprocess.raw_data_functions.get_minor_allele_freq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_minor_allele_freq</span></code></a>(X)</p></td>
<td><p>Compute minor allele frequencies of genotype matrix</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.create_maf_filter" title="preprocess.raw_data_functions.create_maf_filter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_maf_filter</span></code></a>(maf, freq)</p></td>
<td><p>Create minor allele frequency filter</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_datasplit_user_input" title="preprocess.raw_data_functions.check_datasplit_user_input"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_datasplit_user_input</span></code></a>(user_datasplit, user_n_outerfolds, user_n_innerfolds, user_test_set_size_percentage, user_val_set_size_percentage, datasplit, param_to_check)</p></td>
<td><p>Check if user input of data split parameters differs from standard values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.check_train_test_splits" title="preprocess.raw_data_functions.check_train_test_splits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">check_train_test_splits</span></code></a>(y, datasplit, datasplit_params)</p></td>
<td><p>Create stratified train-test splits. Continuous values will be grouped into bins and stratified according to those</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.make_bins" title="preprocess.raw_data_functions.make_bins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_bins</span></code></a>(y, datasplit, datasplit_params)</p></td>
<td><p>Create bins of continuous values for stratification</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.make_nested_cv" title="preprocess.raw_data_functions.make_nested_cv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_nested_cv</span></code></a>(y, outerfolds, innerfolds)</p></td>
<td><p>Create index dictionary for stratified nested cross validation with the following structure:</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.make_stratified_cv" title="preprocess.raw_data_functions.make_stratified_cv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_stratified_cv</span></code></a>(x, y, split_number)</p></td>
<td><p>Create index dictionary for stratified cross-validation with following structure:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#preprocess.raw_data_functions.make_train_test_split" title="preprocess.raw_data_functions.make_train_test_split"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_train_test_split</span></code></a>(y, test_size, val_size=None, val=False, random=42)</p></td>
<td><p>Create index arrays for stratified train-test, respectively train-val-test splits.</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.prepare_data_files">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">prepare_data_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_innerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_encoding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_percentage</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.prepare_data_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare all data files for a common format: genotype matrix, phenotype matrix and index file.</p>
<dl>
<dt>First check if genotype file is .h5 file (standard format of this framework):</dt><dd><dl class="simple">
<dt>YES:    First check if all required information is present in the file, raise Exception if not.</dt><dd><dl class="simple">
<dt>Then check if index file exists:</dt><dd><p>NO: Load genotype and create all required index files
YES: Append all required data splits and maf-filters to index file</p>
</dd>
</dl>
</dd>
</dl>
<p>NO:     Load genotype and create all required files</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – data directory where the phenotype and genotype matrix are stored</p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – name of the genotype matrix including datatype ending</p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – name of the phenotype matrix including datatype ending</p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – name of the phenotype to predict</p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – datasplit to use. Options are: nested-cv, cv-test, train-val-test</p></li>
<li><p><strong>n_outerfolds</strong> (<em>int</em>) – number of outerfolds relevant for nested-cv</p></li>
<li><p><strong>n_innerfolds</strong> (<em>int</em>) – number of folds relevant for nested-cv and cv-test</p></li>
<li><p><strong>test_set_size_percentage</strong> (<em>int</em>) – size of the test set relevant for cv-test and train-val-test</p></li>
<li><p><strong>val_set_size_percentage</strong> (<em>int</em>) – size of the validation set relevant for train-val-test</p></li>
<li><p><strong>models</strong> – models to consider</p></li>
<li><p><strong>user_encoding</strong> (<em>str</em>) – encoding specified by the user</p></li>
<li><p><strong>maf_percentage</strong> (<em>int</em>) – threshold for MAF filter as percentage value</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_genotype_h5_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_genotype_h5_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encodings</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_genotype_h5_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Check .h5 genotype file.
Should contain:</p>
<blockquote>
<div><p>sample_ids: vector with sample names of genotype matrix,
snp_ids: vector with SNP identifiers of genotype matrix,
X_{enc}: (samples x SNPs)-genotype matrix in enc encoding, where enc might refer to:</p>
<blockquote>
<div><p>‘012’: additive (number of minor alleles)
‘raw’: raw (alleles)</p>
</div></blockquote>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – data directory where the phenotype and genotype matrix are stored</p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – name of the phenotype matrix including datatype ending</p></li>
<li><p><strong>encodings</strong> (<em>list</em>) – list of needed encodings</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_index_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_index_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_index_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if index file is available and if the datasets ‘y’, ‘matched_sample_ids’, ‘X_index’, ‘y_index’ and
‘ma_frequency’ exist.
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:param phenotype_matrix_name: name of the phenotype matrix including datatype ending
:param phenotype: name of the phenotype to predict
:return: bool reflecting check result</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.save_all_data_files">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">save_all_data_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_encoding</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_innerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_set_size_percentage</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.save_all_data_files" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Prepare and save all required data files:</dt><dd><ul class="simple">
<li><p>genotype matrix in unified format as .h5 file with,</p></li>
<li><p>phenotype matrix in unified format as .csv file,</p></li>
<li><p>file containing maf filter and data split indices as .h5</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – data directory where the phenotype and genotype matrix are stored</p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – name of the genotype matrix including datatype ending</p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – name of the phenotype matrix including datatype ending</p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – name of the phenotype to predict</p></li>
<li><p><strong>models</strong> – models to consider</p></li>
<li><p><strong>user_encoding</strong> (<em>str</em>) – encoding specified by the user</p></li>
<li><p><strong>maf_percentage</strong> (<em>int</em>) – threshold for MAF filter as percentage value</p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – datasplit to use. Options are: nested-cv, cv-test, train-val-test</p></li>
<li><p><strong>n_outerfolds</strong> (<em>int</em>) – number of outerfolds relevant for nested-cv</p></li>
<li><p><strong>n_innerfolds</strong> (<em>int</em>) – number of folds relevant for nested-cv and cv-test</p></li>
<li><p><strong>test_set_size_percentage</strong> (<em>int</em>) – size of the test set relevant for cv-test and train-val-test</p></li>
<li><p><strong>val_set_size_percentage</strong> (<em>int</em>) – size of the validation set relevant for train-val-test</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_transform_format_genotype_matrix">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_transform_format_genotype_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">models</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_encoding</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_transform_format_genotype_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the format of the specified genotype matrix.
Unified genotype matrix will be saved in subdirectory data and named NAME_OF_GENOTYPE_MATRIX.h5
Unified format of the .h5 file of the genotype matrix required for the further processes:
mandatory:  sample_ids: vector with sample names of genotype matrix,</p>
<blockquote>
<div><p>SNP_ids: vector with SNP identifiers of genotype matrix,
X_{enc}: (samples x SNPs)-genotype matrix in enc encoding, where enc might refer to:</p>
<blockquote>
<div><p>‘012’: additive (number of minor alleles)
‘raw’: raw  (alleles)</p>
</div></blockquote>
</div></blockquote>
<p>optional:   genotype in additional encodings
Accepts .h5, .hdf5, .h5py, .csv, PLINK binary and PLINK files. .h5, .hdf5, .h5py files must satisfy the unified
format. If the genotype matrix contains constant SNPs, those will be removed and a new file will be saved.
Will open .csv, PLINK and binary PLINK files and generate required .h5 format.
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:param models: models to consider
:param user_encoding: encoding specified by the user
:return: genotype matrix (raw encoded if present, 012 encoded otherwise) and sample ids</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>user_encoding</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_genotype_csv_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_genotype_csv_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encodings</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_genotype_csv_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load .csv genotype file. File must have the following structure:
First column must contain the sample ids, the column names should be the SNP ids.
The values should be the genotype matrix either in additive encoding or in raw encoding.
If the genotype is in raw encoding, additive encoding will be calculated.
If genotype is in additive encoding, only this encoding will be returned
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:param encodings: list of needed encodings
:return: sample ids, SNP ids and genotype in additive / raw encoding (if available)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>encodings</strong> (<em>list</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_genotype_binary_plink_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_genotype_binary_plink_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_genotype_binary_plink_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load binary PLINK file, .bim, .fam, .bed files with same prefix need to be in same folder.
Compute additive and raw encoding of genotype
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:return: sample ids, SNP ids and genotype in raw encoding</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_genotype_plink_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_genotype_plink_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_genotype_plink_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load PLINK files, .map and .ped file with same prefix need to be in same folder.
Accepts GENOTYPENAME.ped and GENOTYPENAME.map as input
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:return: sample ids, SNP ids and genotype in raw encoding</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.create_genotype_h5_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">create_genotype_h5_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">snp_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.create_genotype_h5_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save genotype matrix in unified .h5 file.
Structure:</p>
<blockquote>
<div><p>sample_ids
snp_ids
X_raw (or X_012 if X_raw not available)</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – data directory where the phenotype and genotype matrix are stored</p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – name of the genotype matrix including datatype ending</p></li>
<li><p><strong>sample_ids</strong> (<em>numpy.array</em>) – array containing sample ids of genotype data</p></li>
<li><p><strong>snp_ids</strong> (<em>numpy.array</em>) – array containing snp ids of genotype data</p></li>
<li><p><strong>X</strong> (<em>numpy.array</em>) – matrix containing genotype either in raw or in additive encoding</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_and_load_phenotype_matrix">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_and_load_phenotype_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_and_load_phenotype_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Check and load the specified phenotype matrix. Only accept .csv, .pheno, .txt files.
Sample ids need to be in first column, remaining columns should contain phenotypic values
with phenotype name as column name
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param phenotype_matrix_name: name of the phenotype matrix including datatype ending
:param phenotype: name of the phenotype to predict
:return: DataFrame with sample_ids as index and phenotype values as single column without NAN values</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.genotype_phenotype_matching">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">genotype_phenotype_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.genotype_phenotype_matching" title="Permalink to this definition">¶</a></dt>
<dd><p>Match the handed over genotype and phenotype matrix for the phenotype specified by the user
:param X: genotype matrix in additive encoding
:param X_ids: sample ids of genotype matrix
:param y: pd.DataFrame containing sample ids of phenotype as index and phenotype values as single column
:return: matched genotype matrix, matched sample ids, index arrays for genotype and phenotype to redo matching</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>X_ids</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>y</strong> (<em>pandas.DataFrame</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.get_matched_data">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">get_matched_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.get_matched_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Get elements of data specified in index array
:param data: matrix or array
:param index: index array
:return: data at selected indices</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>index</strong> (<em>numpy.array</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.append_index_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">append_index_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_innerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_percentage</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.append_index_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Check index file, described in create_index_file(), and append datasets if necessary
:param data_dir: data directory where the phenotype and genotype matrix are stored
:param genotype_matrix_name: name of the genotype matrix including datatype ending
:param phenotype_matrix_name: name of the phenotype matrix including datatype ending
:param phenotype: name of the phenotype to predict
:param maf_percentage: threshold for MAF filter as percentage value
:param datasplit: datasplit to use. Options are: nested-cv, cv-test, train-val-test
:param n_outerfolds: number of outerfolds relevant for nested-cv
:param n_innerfolds: number of folds relevant for nested-cv and cv-test
:param test_set_size_percentage: size of the test set relevant for cv-test and train-val-test
:param val_set_size_percentage: size of the validation set relevant for train-val-test</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – </p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – </p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – </p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – </p></li>
<li><p><strong>n_outerfolds</strong> (<em>int</em>) – </p></li>
<li><p><strong>n_innerfolds</strong> (<em>int</em>) – </p></li>
<li><p><strong>test_set_size_percentage</strong> (<em>int</em>) – </p></li>
<li><p><strong>val_set_size_percentage</strong> (<em>int</em>) – </p></li>
<li><p><strong>maf_percentage</strong> (<em>int</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.create_index_file">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">create_index_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_matrix_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_innerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.create_index_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Create the .h5 index file containing the maf filters and data splits for the combination of genotype matrix,
phenotype matrix and phenotype.
It will be created using standard values additionally to user inputs for the maf filters and data splits.
Unified format of .h5 file containing the maf filters and data splits:</p>
<blockquote>
<div><dl>
<dt>‘matched_data’: {</dt><dd><p>‘y’: matched phenotypic values,
‘matched_sample_ids’: sample ids of matched genotype/phenotype,
‘X_index’: indices of genotype matrix to redo matching,
‘y_index’: indices of phenotype vector to redo matching,
‘ma_frequency’: minor allele frequency of each SNP in genotype file to create new MAF filters
}</p>
</dd>
<dt>‘maf_filter’: {</dt><dd><p>‘maf_{maf_percentage}’: indices of SNPs to delete (with MAF &lt; maf_percentage),
…
}</p>
</dd>
<dt>‘datasplits’: {</dt><dd><dl>
<dt>‘nested_cv’: {</dt><dd><dl>
<dt>‘#outerfolds-#innerfolds’: {</dt><dd><dl class="simple">
<dt>‘outerfold_0’: {</dt><dd><p>‘innerfold_0’: {‘train’: indices_train, ‘val’: indices_val},
…
‘innerfold_n’: {‘train’: indices_train, ‘val’: indices_val},
‘test’: test_indices
},</p>
</dd>
</dl>
<p>…
‘outerfold_m’: {</p>
<blockquote>
<div><p>‘innerfold_0’: {‘train’: indices_train, ‘val’: indices_val},
…
‘innerfold_n’: {‘train’: indices_train, ‘val’: indices_val},
‘test’: test_indices
}</p>
</div></blockquote>
<p>},</p>
</dd>
</dl>
<p>…
}</p>
</dd>
<dt>‘cv-test’: {</dt><dd><dl>
<dt>‘#folds-test_percentage’: {</dt><dd><dl class="simple">
<dt>‘outerfold_0’: {</dt><dd><p>‘innerfold_0’: {‘train’: indices_train, ‘val’: indices_val},
…
‘innerfold_n’: {‘train’: indices_train, ‘val’: indices_val},
‘test’: test_indices
}</p>
</dd>
</dl>
<p>},</p>
</dd>
</dl>
<p>…
}</p>
</dd>
<dt>‘train-val-test’: {</dt><dd><dl>
<dt>‘train_percentage-val_percentage-test_percentage’: {</dt><dd><dl class="simple">
<dt>‘outerfold_0’: {</dt><dd><p>‘innerfold_0’: {‘train’: indices_train, ‘val’: indices_val},
‘test’: test_indices
}</p>
</dd>
</dl>
<p>},</p>
</dd>
</dl>
<p>…
}</p>
</dd>
</dl>
<p>}</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Standard values for the maf filters and data splits:</dt><dd><p>maf thresholds: 1, 3, 5
folds (inner-/outerfolds for ‘nested-cv’ and folds for ‘cv-test’): 5
test percentage (for ‘cv-test’ and ‘train-val-test’): 20
val percentage (for ‘train-val-test’): 20</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_dir</strong> (<em>str</em>) – data directory where the phenotype and genotype matrix are stored</p></li>
<li><p><strong>genotype_matrix_name</strong> (<em>str</em>) – name of the genotype matrix including datatype ending</p></li>
<li><p><strong>phenotype_matrix_name</strong> (<em>str</em>) – name of the phenotype matrix including datatype ending</p></li>
<li><p><strong>phenotype</strong> (<em>str</em>) – name of the phenotype to predict</p></li>
<li><p><strong>maf_percentage</strong> (<em>int</em>) – threshold for MAF filter as percentage value</p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – datasplit to use. Options are: nested-cv, cv-test, train-val-test</p></li>
<li><p><strong>n_outerfolds</strong> (<em>int</em>) – number of outerfolds relevant for nested-cv</p></li>
<li><p><strong>n_innerfolds</strong> (<em>int</em>) – number of folds relevant for nested-cv and cv-test</p></li>
<li><p><strong>test_set_size_percentage</strong> (<em>int</em>) – size of the test set relevant for cv-test and train-val-test</p></li>
<li><p><strong>val_set_size_percentage</strong> (<em>int</em>) – size of the validation set relevant for train-val-test</p></li>
<li><p><strong>X</strong> (<em>numpy.array</em>) – genotype in additive encoding to create ma-frequencies</p></li>
<li><p><strong>y</strong> (<em>numpy.array</em>) – matched phenotype values</p></li>
<li><p><strong>sample_ids</strong> (<em>numpy.array</em>) – matched sample ids of genotype/phenotype</p></li>
<li><p><strong>X_index</strong> (<em>numpy.array</em>) – index file of genotype to redo matching</p></li>
<li><p><strong>y_index</strong> (<em>numpy.array</em>) – index file of phenotype to redo matching</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.filter_non_informative_snps">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">filter_non_informative_snps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.filter_non_informative_snps" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove non-informative SNPs, i.e. SNPs that are constant
:param X: genotype matrix in raw or additive encoding
:return: filtered genotype matrix and filter-vector</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>X</strong> (<em>numpy.array</em>) – </p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.get_minor_allele_freq">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">get_minor_allele_freq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.get_minor_allele_freq" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute minor allele frequencies of genotype matrix
:param X: genotype matrix in additive encoding
:return: array with frequencies</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>X</strong> (<em>numpy.array</em>) – </p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.create_maf_filter">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">create_maf_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">maf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">freq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.create_maf_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Create minor allele frequency filter
:param maf: maf threshold as percentage value
:param freq: array containing minor allele frequencies as decimal value
:return: array containing indices of SNPs with MAF smaller than specified threshold, i.e. SNPs to delete</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>maf</strong> (<em>int</em>) – </p></li>
<li><p><strong>freq</strong> (<em>numpy.array</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_datasplit_user_input">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_datasplit_user_input</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">user_datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_n_outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_n_innerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_test_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_val_set_size_percentage</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param_to_check</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_datasplit_user_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if user input of data split parameters differs from standard values.
If it does, add input to list of parameters
:param user_datasplit: datasplit specified by the user
:param user_n_outerfolds: number of outerfolds relevant for nested-cv specified by the user
:param user_n_innerfolds: number of folds relevant for nested-cv and cv-test specified by the user
:param user_test_set_size_percentage:</p>
<blockquote>
<div><p>size of the test set relevant for cv-test and train-val-test specified by the user</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>user_val_set_size_percentage</strong> (<em>int</em>) – size of the validation set relevant for train-val-test specified by the user</p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – type of data split</p></li>
<li><p><strong>param_to_check</strong> (<em>list</em>) – standard parameters to compare to</p></li>
<li><p><strong>user_datasplit</strong> (<em>str</em>) – </p></li>
<li><p><strong>user_n_outerfolds</strong> (<em>int</em>) – </p></li>
<li><p><strong>user_n_innerfolds</strong> (<em>int</em>) – </p></li>
<li><p><strong>user_test_set_size_percentage</strong> (<em>int</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>adapted list of parameters</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.check_train_test_splits">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">check_train_test_splits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit_params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.check_train_test_splits" title="Permalink to this definition">¶</a></dt>
<dd><p>Create stratified train-test splits. Continuous values will be grouped into bins and stratified according to those
:param datasplit: type of datasplit (‘nested-cv’, ‘cv-test’, ‘train-val-test’)
:param y: array with phenotypic values for stratification
:param datasplit_params: parameters to use for split:</p>
<blockquote>
<div><p>[n_outerfolds, n_innerfolds] for nested-cv
[n_innerfolds, test_set_size_percentage] for cv-test
[val_set_size_percentage, test_set_size_percentage] for train-val-test</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dictionary respectively arrays with indices</p>
</dd>
<dt class="field-even">Parameters</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>y</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – </p></li>
<li><p><strong>datasplit_params</strong> (<em>list</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.make_bins">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">make_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">datasplit_params</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.make_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Create bins of continuous values for stratification
:param y: array containing phenotypic values
:param datasplit: train test split to use
:param datasplit_params: list of parameters to use:</p>
<blockquote>
<div><p>[n_outerfolds, n_innerfolds] for nested-cv
[n_innerfolds, test_set_size_percentage] for cv-test
[val_set_size_percentage, test_set_size_percentage] for train-val-test</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>binned array</p>
</dd>
<dt class="field-even">Parameters</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>y</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>datasplit</strong> (<em>str</em>) – </p></li>
<li><p><strong>datasplit_params</strong> (<em>list</em>) – </p></li>
</ul>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.make_nested_cv">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">make_nested_cv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outerfolds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">innerfolds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.make_nested_cv" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Create index dictionary for stratified nested cross validation with the following structure:</dt><dd><p>{‘outerfold_0_test’: test_indices,
‘outerfold_0’: {fold_0_train: innerfold_0_train_indices,</p>
<blockquote>
<div><p>fold_0_test: innerfold_0_test_indices,
…
fold_n_train: innerfold_n_train_indices,
fold_n_test: innerfold_n_test_indices
},</p>
</div></blockquote>
<p>…
‘outerfold_m_test’: test_indices,
‘outerfold_m’: {fold_0_train: innerfold_0_train_indices,</p>
<blockquote>
<div><p>fold_0_test: innerfold_0_test_indices,
…
fold_n_train: innerfold_n_train_indices,
fold_n_test: innerfold_n_test_indices
}</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y</strong> (<em>numpy.array</em>) – target values grouped in bins for stratification</p></li>
<li><p><strong>outerfolds</strong> (<em>int</em>) – number of outer folds</p></li>
<li><p><strong>innerfolds</strong> (<em>int</em>) – number of inner folds</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index dictionary</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.make_stratified_cv">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">make_stratified_cv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_number</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.make_stratified_cv" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Create index dictionary for stratified cross-validation with following structure:</dt><dd><p>{fold_0_train: fold_0_train_indices,
fold_0_test: fold_0_test_indices,
…
fold_n_train: fold_n_train_indices,
fold_n_test: fold_n_test_indices
}</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>numpy.array</em>) – whole train indices</p></li>
<li><p><strong>y</strong> (<em>numpy.array</em>) – target values binned in groups for stratification</p></li>
<li><p><strong>split_number</strong> (<em>int</em>) – number of folds</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary containing train and validation indices for each fold</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="preprocess.raw_data_functions.make_train_test_split">
<span class="sig-prename descclassname"><span class="pre">preprocess.raw_data_functions.</span></span><span class="sig-name descname"><span class="pre">make_train_test_split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">42</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#preprocess.raw_data_functions.make_train_test_split" title="Permalink to this definition">¶</a></dt>
<dd><p>Create index arrays for stratified train-test, respectively train-val-test splits.
:param y: target values grouped in bins for stratification
:param test_size: size of test set as percentage value
:param val_size: size of validation set as percentage value
:param val: if True, function returns validation set additionally to train and test set
:param random: controls shuffling of data
:return: either train, val and test index arrays or</p>
<blockquote>
<div><p>train and test index arrays and corresponding binned target values</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y</strong> (<em>numpy.array</em>) – </p></li>
<li><p><strong>test_size</strong> (<em>int</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(numpy.array, numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="../../../index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">preprocess.raw_data_functions</span></code></a><ul>
<li><a class="reference internal" href="#module-contents">Module Contents</a><ul>
<li><a class="reference internal" href="#functions">Functions</a><ul>
<li><a class="reference internal" href="#preprocess.raw_data_functions.prepare_data_files">prepare_data_files</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_h5_file">check_genotype_h5_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_index_file">check_index_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.save_all_data_files">save_all_data_files</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_transform_format_genotype_matrix">check_transform_format_genotype_matrix</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_csv_file">check_genotype_csv_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_binary_plink_file">check_genotype_binary_plink_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_genotype_plink_file">check_genotype_plink_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.create_genotype_h5_file">create_genotype_h5_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_and_load_phenotype_matrix">check_and_load_phenotype_matrix</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.genotype_phenotype_matching">genotype_phenotype_matching</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.get_matched_data">get_matched_data</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.append_index_file">append_index_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.create_index_file">create_index_file</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.filter_non_informative_snps">filter_non_informative_snps</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.get_minor_allele_freq">get_minor_allele_freq</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.create_maf_filter">create_maf_filter</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_datasplit_user_input">check_datasplit_user_input</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.check_train_test_splits">check_train_test_splits</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.make_bins">make_bins</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.make_nested_cv">make_nested_cv</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.make_stratified_cv">make_stratified_cv</a></li>
<li><a class="reference internal" href="#preprocess.raw_data_functions.make_train_test_split">make_train_test_split</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../../_sources/autoapi/preprocess/raw_data_functions/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">easyPheno  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href=""><code class="xref py py-mod docutils literal notranslate"><span class="pre">preprocess.raw_data_functions</span></code></a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2022, Florian Haselbeck, Maura John, Dominik G. Grimm.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.4.0.
    </div>
  </body>
</html>